### PR TITLE
[WIP] Add basic ExtensionIndex class

### LINKED ISSUE #22861: Add interface for defining an ExtensionIndex
These occur naturally with `.groupby(extensionarray)` and `Seres[extensionarray].value_counts`. We should define a public API so that we don't have to convert to object in these cases.

You'd likely end up with an `Index` with a non-object dtype.

---

*more radically* this kind of removes the *need* for all our Index subclasses, aside from MultiIndex. But we can think about that separately from the interface.

I don't think this is blocking for 1.0, but it may not be too much effort.

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
Explored this a bit 2 weeks ago, so thought could open it as a WIP PR in case it might serve discussion.

For me, the main question is how "complete" we want this to be, before we consider merging it. This WIP certainly already is able to preserve the EA in the Index (not convert to objects), and basic indexing works (but only tested the basics, and not all combinations of uniques/duplicated, sorted/non-sorted ... indexes)

Closes https://github.com/pandas-dev/pandas/issues/22861

### CODE REVIEW

File: pandas/core/indexes/extension.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        """ return the dtype str of the underlying data """
+        return str(self.dtype)
+
+    @property
+    def _values(self):
+        return self._data
+
+    @property
+    def values(self):
+        """ return the underlying data as an ndarray """
```
Comment (TomAugspurger) [inline pandas/core/indexes/extension.py]: ndarray -> extension array.

**Comment by TomAugspurger** [comment]
Interesting, this is less work than I expected. If I were to do `df.groupby('extension_array').sum()`, would we get back an extension index? Or do we need to update the `Index.__new__` to make an `ExtensionIndex` in that case?

**Comment by TomAugspurger** [comment]
Ah, I completely overlooked that.

**Comment by jorisvandenbossche** [comment]
Putting here a comment I already wrote a time ago, but the actual discussion can maybe go in https://github.com/pandas-dev/pandas/issues/23565

The main question here is: **which values to use for indexing?**

Currently, in this PR I tried to create the correct indexing Engine based on the values of `EA._values_for_factorize`, because that seemed a natural fit (for indexing, they need to be hashable like for factorize). 

However, in many cases, we are using `_ndarray_values` internally. And eg for IntegerArray, those two are actually different: `_values_for_factorize` is an object array (to have the NaNs), but `_ndarray_values` is the integers. This raises errors in the engine not expecting an integer array. Of course, we might need to change `_ndarray_values` for `IntegerArray` to also return object, but that also means that indexing would be quite expensive in general (using the object dtype path)

**Comment by TomAugspurger** [comment]
Just thinking out loud here: For IntegerArray, one could in principle write an engine that uses `_ndarray_values`, correct? You would need to take care to mask the "missing values" (1s) in every operation, but it should be doable.

Regardless, I'm ok with a default "ExtensionIndex" (an Index with the extension `dtype` preserved) for now, with the possibility of more customizable extension index classes later.

**Comment by jorisvandenbossche** [comment]
> Just thinking out loud here: For IntegerArray, one could in principle write an engine that uses _ndarray_values, correct? You would need to take care to mask the "missing values" (1s) in every operation, but it should be doable.

But that would mean writing a custom engine?

Anyway (also future wise), an engine that works with values/mask combination might be useful in general. 
But that is not for this PR :-) 

I suppose short term the easiest is to use the object dtype values for IntegerArray for all indexing related things.

**Comment by TomAugspurger** [comment]
> But that would mean writing a custom engine?

Right, it would push more work onto the EA author (which is us for IntegerArray). Pandas would provide the hooks somewhere in `pd.Index.__new__`

```python
engine = getattr(ExtensionArray, '_index_engine', default_engine)
```

> I suppose short term the easiest is to use the object dtype values for IntegerArray for all indexing related things.


Agreed. Ensuring that `Index.dtype` is `Int64Dtype`, and not object, seems the most important thing. Then we can give speed improvements without breaking API.

**Comment by jreback** [comment]
nice idea for 0.25 :>

**Comment by jreback** [comment]
good idea, but needs a reboot.

**Comment by Detry322** [comment]
Still hoping for this feature!

**Comment by Melendowski** [comment]
I'm wondering what the current thoughts on this issue are wrt to 

https://github.com/pandas-dev/pandas/issues/30001

Is it being considered a non issue or a different work around being put in place?

**Comment by jreback** [comment]
see #22681

**Pull request closed without merging**